File: 苦工ai-agnes/scripts/agnes_helper.py
"""
Agnes API 苦工 — Python 版
带 Key Pool 自动轮换 + 文本/图片/视频 全套能力
Key来源（优先级）:
  1. config.local.json（与本脚本同目录）← 推荐
  2. E:\\ai\\vedio_maker\\config.local.json（项目旧路径）
  3. 环境变量 AGNES_KEYS（JSON数组字符串）
"""

import json
import os
import time
import urllib.request
import urllib.error

BASE = "https://apihub.agnes-ai.com/v1"


class Agnes:
    """Agnes API 苦工：免费的多模态AI工具"""

    def __init__(self, config_path: str = None):
        self._script_dir = os.path.dirname(os.path.abspath(__file__))
        self.config_path = config_path  # 如果传了则优先
        self._key_pool = self._load_keys()
        self._key_index = 0
        self._failed_keys = {}  # key -> timestamp

    # ========== Key Pool ==========

    def _load_keys(self):
        # 方式0: 显式传入路径
        if self.config_path:
            try:
                with open(self.config_path, "r", encoding="utf-8-sig") as f:
                    keys = json.load(f).get("agnes_keys", [])
                    if keys: return keys
            except Exception:
                pass

        # 方式1: 本目录的 config.local.json
        local_path = os.path.join(self._script_dir, "config.local.json")
        try:
            with open(local_path, "r", encoding="utf-8-sig") as f:
                keys = json.load(f).get("agnes_keys", [])
                if keys: return keys
        except Exception:
            pass

        # 方式2: 项目旧路径
        old_path = r"E:\ai\vedio_maker\config.local.json"
        try:
            with open(old_path, "r", encoding="utf-8-sig") as f:
                keys = json.load(f).get("agnes_keys", [])
                if keys: return keys
        except Exception:
            pass

        # 方式3: 环境变量
        env = os.environ.get("AGNES_KEYS")
        if env:
            try:
                keys = json.loads(env)
                if keys: return keys
            except Exception:
                pass

        # 都没找到
        raise RuntimeError(
            "❌ 未找到 Agnes API Key！\n"
            f"请创建 {local_path}，内容格式见 config.example.json"
        )

    def _get_next_key(self):
        now = time.time()
        for _ in range(len(self._key_pool)):
            idx = self._key_index % len(self._key_pool)
            self._key_index += 1
            k = self._key_pool[idx]
            if k in self._failed_keys and (now - self._failed_keys[k]) < 60:
                continue
            return k
        self._failed_keys = {}
        return self._key_pool[0]

    def _fetch(self, endpoint: str, body: dict, timeout_ms: int = 60000) -> dict:
        api_key = self._get_next_key()
        data_bytes = json.dumps(body).encode("utf-8")
        req = urllib.request.Request(
            BASE + endpoint,
            data=data_bytes,
            headers={
                "Authorization": f"Bearer {api_key}",
                "Content-Type": "application/json",
            },
            method="POST",
        )
        try:
            resp = urllib.request.urlopen(req, timeout=timeout_ms / 1000)
            result = json.loads(resp.read().decode("utf-8"))
            if "error" in result and result["error"]:
                raise Exception(result["error"] if isinstance(result["error"], str) else result["error"].get("message", str(result["error"])))
            return result
        except Exception as e:
            self._failed_keys[api_key] = time.time()
            raise e

    def _get(self, url: str, timeout_ms: int = 30000) -> dict:
        """发起 GET 请求（用于查询视频结果）"""
        api_key = self._get_next_key()
        req = urllib.request.Request(url, headers={"Authorization": f"Bearer {api_key}"})
        try:
            resp = urllib.request.urlopen(req, timeout=timeout_ms / 1000)
            return json.loads(resp.read().decode("utf-8"))
        except Exception as e:
            self._failed_keys[api_key] = time.time()
            raise e

    # ========== 1. 文本能力 ==========

    def chat(self, prompt: str, system_prompt: str = None, **kwargs) -> str:
        """单条文本对话"""
        messages = []
        if system_prompt:
            messages.append({"role": "system", "content": system_prompt})
        messages.append({"role": "user", "content": prompt})

        body = {
            "model": "agnes-2.0-flash",
            "messages": messages,
            "temperature": kwargs.get("temperature", 0.7),
            "max_tokens": kwargs.get("max_tokens", 4096),
        }
        if kwargs.get("response_format") == "json":
            body["response_format"] = {"type": "json_object"}

        data = self._fetch("/chat/completions", body, kwargs.get("timeout_ms", 60000))
        return data["choices"][0]["message"]["content"]

    def batch(self, prompts: list, delay_ms: int = 1000) -> list:
        """批量文本调用：prompts = [(prompt, system_prompt), ...]"""
        results = []
        for i, (prompt, system_prompt) in enumerate(prompts):
            r = self.chat(prompt, system_prompt)
            results.append(r)
            if i < len(prompts) - 1:
                time.sleep(delay_ms / 1000)
        return results

    def format_json(self, raw_data: str) -> list:
        """将原始数据格式化为JSON"""
        result = self.chat(
            f"将以下数据格式化为标准JSON数组:\n{raw_data}",
            "只输出合法JSON，不要markdown包裹",
            response_format="json",
            temperature=0.3,
        )
        return json.loads(result)

    @staticmethod
    def validate_json(json_str: str, schema: dict = None) -> dict:
        """校验JSON格式和字段完整性"""
        errors = []
        schema = schema or {}
        try:
            data = json.loads(json_str)
            arr = data if isinstance(data, list) else [data]
            min_count = schema.get("minCount")
            if min_count and len(arr) < min_count:
                errors.append(f"数量不足: 需要{min_count}条, 实际{len(arr)}条")
            fields = schema.get("requiredFields", ["id", "name", "url"])
            for i, item in enumerate(arr):
                for f in fields:
                    if f not in item or not item[f]:
                        errors.append(f"第{i + 1}条缺字段: {f}")
            return {"pass": len(errors) == 0, "errors": errors}
        except json.JSONDecodeError as e:
            return {"pass": False, "errors": [f"JSON解析失败: {e}"]}

    # ========== 2. 图片能力 ==========

    def image(self, prompt: str, size: str = "1024x768", **kwargs) -> str:
        """文生图，返回图片URL或Base64"""
        body = {
            "model": "agnes-image-2.1-flash",
            "prompt": prompt,
            "size": size,
        }
        if kwargs.get("return_base64"):
            body["return_base64"] = True
        else:
            body["extra_body"] = {"response_format": "url"}

        data = self._fetch("/images/generations", body, kwargs.get("timeout_ms", 120000))
        img = data["data"][0]
        if kwargs.get("return_base64"):
            return img["b64_json"]
        return img.get("url") or img.get("b64_json")

    def image_edit(self, prompt: str, image_url: str, size: str = "1024x768", **kwargs) -> str:
        """图生图（风格迁移/编辑），返回图片URL"""
        body = {
            "model": "agnes-image-2.1-flash",
            "prompt": prompt,
            "size": size,
            "extra_body": {
                "image": [image_url],
                "response_format": "url",
            },
        }
        data = self._fetch("/images/generations", body, kwargs.get("timeout_ms", 120000))
        return data["data"][0]["url"]

    # ========== 3. 视频能力 ==========

    def video_submit(self, prompt: str, image_url: str = None, **kwargs) -> dict:
        """提交视频生成任务，不等待"""
        body = {
            "model": "agnes-video-v2.0",
            "prompt": prompt,
        }
        if image_url:
            body["image"] = image_url
        for k in ("num_frames", "frame_rate", "width", "height", "negative_prompt", "seed"):
            if k in kwargs:
                body[k] = kwargs[k]

        return self._fetch("/videos", body, kwargs.get("timeout_ms", 30000))

    def video_query(self, video_id: str) -> dict:
        """查询视频生成结果（推荐用 video_id）"""
        url = f"https://apihub.agnes-ai.com/agnesapi?video_id={video_id}"
        return self._get(url, 30000)

    def video(self, prompt: str, **kwargs) -> dict:
        """文生视频（提交+自动轮询等待）"""
        return self._video_submit_and_poll(prompt, None, **kwargs)

    def video_from_image(self, prompt: str, image_url: str, **kwargs) -> dict:
        """图生视频（提交+自动轮询等待）"""
        return self._video_submit_and_poll(prompt, image_url, **kwargs)

    def _video_submit_and_poll(self, prompt: str, image_url: str = None, **kwargs) -> dict:
        """内部：提交视频任务+轮询"""
        wait = kwargs.get("wait", True)
        poll_interval = kwargs.get("poll_interval", 5)
        timeout_ms = kwargs.get("timeout_ms", 300000)

        task = self.video_submit(prompt, image_url, **kwargs)
        video_id = task.get("video_id")
        task_id = task.get("task_id") or task.get("id")

        if not wait:
            return {"task_id": task_id, "video_id": video_id, "status": "queued"}

        deadline = time.time() + timeout_ms / 1000
        while time.time() < deadline:
            time.sleep(poll_interval)
            result = self.video_query(video_id)
            status = result.get("status")
            if status == "completed":
                return {
                    "task_id": task_id,
                    "video_id": video_id,
                    "video_url": result.get("remixed_from_video_id") or result.get("video_url"),
                    "seconds": result.get("seconds"),
                    "status": "completed",
                    "size": result.get("size"),
                }
            if status == "failed":
                raise Exception(f"视频生成失败: {result.get('error', '未知错误')}")

        raise Exception(f"视频生成超时 ({timeout_ms}ms)")

    @property
    def key_count(self) -> int:
        return len(self._key_pool)

File: 苦工ai-agnes/scripts/agnes_helper.js
/**
 * Agnes API 苦工 — 带 Key Pool 自动轮换 + 文本/图片/视频 全套
 * 
 * 9个Key自动轮换，rate limit自动换下一个，60秒后重试
 * Key来源（优先级）:
 *   1. config.local.json（与本脚本同目录）← 推荐
 *   2. E:\ai\vedio_maker\config.local.json（项目旧路径）
 *   3. 环境变量 AGNES_KEYS（JSON数组字符串）
 * API文档: https://agnes-ai.com/doc/overview
 */

const fs = require("fs");
const path = require("path");

// ========== Key Pool ==========

function loadKeys() {
  // 方式1: 本目录的 config.local.json
  var localPath = path.join(__dirname, "config.local.json");
  try {
    var r = fs.readFileSync(localPath, "utf-8").replace(/^\uFEFF/, "");
    var keys = JSON.parse(r).agnes_keys;
    if (keys && keys.length > 0) return keys;
  } catch (e) { /* fall through */ }

  // 方式2: 项目旧路径
  var oldPath = "E:\\ai\\vedio_maker\\config.local.json";
  try {
    var r = fs.readFileSync(oldPath, "utf-8").replace(/^\uFEFF/, "");
    var keys = JSON.parse(r).agnes_keys;
    if (keys && keys.length > 0) return keys;
  } catch (e) { /* fall through */ }

  // 方式3: 环境变量
  var env = process.env.AGNES_KEYS;
  if (env) {
    try { var keys = JSON.parse(env); if (keys && keys.length > 0) return keys; } catch (e) {}
  }

  // 都没找到 — 报错
  throw new Error(
    "❌ 未找到 Agnes API Key！\n" +
    "请创建 " + localPath + "，内容格式见 config.example.json"
  );
}

var keyIndex = 0;
var keyPool = loadKeys();
var failedKeys = {};

function getNextKey() {
  var now = Date.now();
  for (var i = 0; i < keyPool.length; i++) {
    var idx = keyIndex % keyPool.length;
    keyIndex++;
    var k = keyPool[idx];
    if (failedKeys[k] && now - failedKeys[k] < 60000) continue;
    return k;
  }
  failedKeys = {};
  return keyPool[0];
}

var BASE = "https://apihub.agnes-ai.com/v1";

async function agnesFetch(endpoint, body, timeoutMs) {
  timeoutMs = timeoutMs || 60000;
  var apiKey = getNextKey();
  var controller = new AbortController();
  var timer = setTimeout(function () { controller.abort(); }, timeoutMs);
  try {
    var resp = await fetch(BASE + endpoint, {
      method: "POST",
      headers: { Authorization: "Bearer " + apiKey, "Content-Type": "application/json" },
      body: JSON.stringify(body),
      signal: controller.signal,
    });
    clearTimeout(timer);
    var data = await resp.json();
    if (data.error) throw new Error(typeof data.error === "string" ? data.error : data.error.message);
    return data;
  } catch (e) {
    failedKeys[apiKey] = Date.now();
    clearTimeout(timer);
    throw e;
  }
}

// ========== 1. 文本能力 ==========

async function agnesChat(prompt, systemPrompt, options) {
  options = options || {};
  var messages = [];
  if (systemPrompt) messages.push({ role: "system", content: systemPrompt });
  messages.push({ role: "user", content: prompt });
  var body = {
    model: "agnes-2.0-flash",
    messages: messages,
    temperature: options.temperature || 0.7,
    max_tokens: options.max_tokens || 4096,
  };
  if (options.response_format === "json") body.response_format = { type: "json_object" };
  var data = await agnesFetch("/chat/completions", body, options.timeoutMs);
  return data.choices[0].message.content;
}

async function agnesBatch(prompts, delayMs) {
  delayMs = delayMs || 1000;
  var results = [];
  for (var i = 0; i < prompts.length; i++) {
    var p = prompts[i];
    var r = await agnesChat(p.prompt, p.systemPrompt, p.options);
    results.push(r);
    if (i < prompts.length - 1) await new Promise(function (r2) { setTimeout(r2, delayMs); });
  }
  return results;
}

async function formatToolsJson(rawData) {
  var result = await agnesChat("将以下工具数据格式化为标准JSON数组:\n" + rawData, "只输出合法JSON，不要markdown包裹", {
    response_format: "json",
    temperature: 0.3,
  });
  return JSON.parse(result);
}

function validateJson(jsonStr, schema) {
  var errors = [];
  schema = schema || {};
  try {
    var data = JSON.parse(jsonStr);
    var arr = Array.isArray(data) ? data : [data];
    if (schema.minCount && arr.length < schema.minCount)
      errors.push("数量不足: 需要" + schema.minCount + "条, 实际" + arr.length + "条");
    var fields = schema.requiredFields || ["id", "name", "url"];
    for (var i = 0; i < arr.length; i++)
      for (var f = 0; f < fields.length; f++)
        if (!arr[i][fields[f]]) errors.push("第" + (i + 1) + "条缺字段: " + fields[f]);
    return { pass: errors.length === 0, errors: errors };
  } catch (e) {
    return { pass: false, errors: ["JSON解析失败: " + e.message] };
  }
}

// ========== 2. 图片能力 ==========

/**
 * 文生图
 * @param {string} prompt - 图片描述
 * @param {string} size - 尺寸如 "1024x768"
 * @param {object} options - { returnBase64: bool, timeoutMs: number }
 * @returns {string} 图片URL 或 Base64 字符串
 */
async function agnesImage(prompt, size, options) {
  options = options || {};
  var body = {
    model: "agnes-image-2.1-flash",
    prompt: prompt,
    size: size || "1024x768",
  };
  if (options.returnBase64) {
    body.return_base64 = true;
  } else {
    body.extra_body = { response_format: "url" };
  }
  var data = await agnesFetch("/images/generations", body, options.timeoutMs || 120000);
  var img = data.data[0];
  if (options.returnBase64) return img.b64_json;
  return img.url || img.b64_json;
}

/**
 * 图生图（风格迁移/编辑）
 * @param {string} prompt - 编辑描述
 * @param {string} imageUrl - 公网可访问的图片URL
 * @param {string} size - 尺寸
 * @param {object} options
 * @returns {string} 编辑后的图片URL
 */
async function agnesImageEdit(prompt, imageUrl, size, options) {
  options = options || {};
  var body = {
    model: "agnes-image-2.1-flash",
    prompt: prompt,
    size: size || "1024x768",
    extra_body: {
      image: [imageUrl],
      response_format: "url",
    },
  };
  var data = await agnesFetch("/images/generations", body, options.timeoutMs || 120000);
  return data.data[0].url;
}

// ========== 3. 视频能力 ==========

/**
 * 文生视频（提交 + 自动轮询）
 * @param {string} prompt - 视频描述
 * @param {object} options - { num_frames, frame_rate, width, height, wait, pollInterval, timeoutMs }
 * @returns {object} { task_id, video_id, videoUrl, seconds, status }
 */
async function agnesVideo(prompt, options) {
  options = options || {};
  return await agnesVideoSubmitAndPoll(prompt, null, options);
}

/**
 * 图生视频（提交 + 自动轮询）
 * @param {string} prompt - 视频描述
 * @param {string} imageUrl - 公网图片URL
 * @param {object} options
 * @returns {object}
 */
async function agnesVideoFromImage(prompt, imageUrl, options) {
  options = options || {};
  return await agnesVideoSubmitAndPoll(prompt, imageUrl, options);
}

async function agnesVideoSubmitAndPoll(prompt, imageUrl, options) {
  options = options || {};
  var wait = options.wait !== false;
  var pollInterval = options.pollInterval || 5000;
  var timeoutMs = options.timeoutMs || 300000;

  // 提交任务
  var task = await agnesVideoSubmit(prompt, imageUrl, options);
  var videoId = task.video_id;
  var taskId = task.task_id || task.id;

  if (!wait) return { task_id: taskId, video_id: videoId, status: "queued" };

  // 轮询等待
  var deadline = Date.now() + timeoutMs;
  while (Date.now() < deadline) {
    await new Promise(function (r) { setTimeout(r, pollInterval); });
    var result = await agnesVideoQuery(videoId);
    if (result.status === "completed") {
      return {
        task_id: taskId,
        video_id: videoId,
        videoUrl: result.remixed_from_video_id || result.video_url,
        seconds: result.seconds,
        status: "completed",
        size: result.size,
      };
    }
    if (result.status === "failed") {
      throw new Error("视频生成失败: " + (result.error ? JSON.stringify(result.error) : "未知错误"));
    }
  }
  throw new Error("视频生成超时 (" + timeoutMs + "ms)");
}

/**
 * 提交视频任务（不等待）
 * @param {string} prompt
 * @param {string|null} imageUrl
 * @param {object} options - { num_frames, frame_rate, width, height }
 * @returns {object} { id, task_id, video_id, status }
 */
async function agnesVideoSubmit(prompt, imageUrl, options) {
  options = options || {};
  var body = {
    model: "agnes-video-v2.0",
    prompt: prompt,
  };
  if (options.num_frames) body.num_frames = options.num_frames;
  if (options.frame_rate) body.frame_rate = options.frame_rate;
  if (options.width) body.width = options.width;
  if (options.height) body.height = options.height;
  if (imageUrl) body.image = imageUrl;

  return await agnesFetch("/videos", body, options.timeoutMs || 30000);
}

/**
 * 查询视频结果（推荐用 video_id）
 * @param {string} videoId
 * @returns {object} { status, video_url, remixed_from_video_id, seconds, size, error }
 */
async function agnesVideoQuery(videoId) {
  var apiKey = getNextKey();
  try {
    var resp = await fetch("https://apihub.agnes-ai.com/agnesapi?video_id=" + encodeURIComponent(videoId), {
      method: "GET",
      headers: { Authorization: "Bearer " + apiKey },
    });
    return await resp.json();
  } catch (e) {
    failedKeys[apiKey] = Date.now();
    throw e;
  }
}

module.exports = {
  // 文本
  agnesChat: agnesChat,
  agnesBatch: agnesBatch,
  formatToolsJson: formatToolsJson,
  validateJson: validateJson,
  // 图片
  agnesImage: agnesImage,
  agnesImageEdit: agnesImageEdit,
  // 视频
  agnesVideo: agnesVideo,
  agnesVideoFromImage: agnesVideoFromImage,
  agnesVideoSubmit: agnesVideoSubmit,
  agnesVideoQuery: agnesVideoQuery,
  // 工具
  keyCount: keyPool.length,
  getKeyPool: function () { return keyPool.length; },
};
